### PR TITLE
1311664: Ignore empty guest IDs and hypervisor IDs

### PR DESCRIPTION
**Pull request by cnsnyder**

### CODE REVIEW

File: server/src/main/java/org/candlepin/resource/HypervisorResource.java [hunk 1]
```diff
@@ -108,7 +110,7 @@ public HypervisorResource(ConsumerResource consumerResource,
     @Deprecated
     @Transactional
     public HypervisorCheckInResult hypervisorUpdate(
-        Map<String, List<GuestId>> hostGuestMap, @Context Principal principal,
+        Map<String, ArrayList<GuestId>> hostGuestMap, @Context Principal principal,
```
Comment (wottop) [inline server/src/main/java/org/candlepin/resource/HypervisorResource.java]: Why the change to ArrayList? List still works.

Comment (cnsnyder) [inline server/src/main/java/org/candlepin/resource/HypervisorResource.java]: I've changed to ArrayList as the List class does not allow removal of items in the list using an iterator (as done below).

File: server/src/main/java/org/candlepin/resource/HypervisorResource.java [hunk 2]
```diff
@@ -123,22 +125,43 @@ public HypervisorCheckInResult hypervisorUpdate(
...
         VirtConsumerMap hypervisorConsumersMap =
                 consumerCurator.getHostConsumersMap(owner, hostGuestMap.keySet());
 
+        int emptyGuestIdCount = 0;
+        int emptyHypervisorIdCount = 0;
         Set<String> allGuestIds = new HashSet<String>();
-        for (Entry<String, List<GuestId>> hostEntry : hostGuestMap.entrySet()) {
-            for (GuestId gid : hostEntry.getValue()) {
-                allGuestIds.add(gid.getGuestId());
+        Set<GuestId> toRemove = new HashSet<GuestId>();
```
Comment (wottop) [inline server/src/main/java/org/candlepin/resource/HypervisorResource.java]: Doesn't get used anywhere in the diff. Still needed?

File: server/src/main/java/org/candlepin/resource/HypervisorResource.java [hunk 3]
```diff
@@ -123,22 +125,43 @@ public HypervisorCheckInResult hypervisorUpdate(
...
+        if (emptyGuestIdCount > 0) {
+            log.warn("Found " + emptyGuestIdCount + " empty guest id(s). Ignoring.");
+        }
         // Maps virt guest ID to registered consumer for guest, if one exists:
         VirtConsumerMap guestConsumersMap = consumerCurator.getGuestConsumersMap(
                 owner, allGuestIds);
 
         HypervisorCheckInResult result = new HypervisorCheckInResult();
-        for (Entry<String, List<GuestId>> hostEntry : hostGuestMap.entrySet()) {
+        for (Entry<String, ArrayList<GuestId>> hostEntry : hostGuestMap.entrySet()) {
```
Comment (wottop) [inline server/src/main/java/org/candlepin/resource/HypervisorResource.java]: Same LIst/ArrayList comment as before

File: server/src/main/java/org/candlepin/resource/HypervisorResource.java [hunk 4]
```diff
@@ -123,22 +125,42 @@ public HypervisorCheckInResult hypervisorUpdate(
...
                 consumerCurator.getHostConsumersMap(owner, hostGuestMap.keySet());
 
+        int emptyGuestIdCount = 0;
+        int emptyHypervisorIdCount = 0;
         Set<String> allGuestIds = new HashSet<String>();
-        for (Entry<String, List<GuestId>> hostEntry : hostGuestMap.entrySet()) {
-            for (GuestId gid : hostEntry.getValue()) {
-                allGuestIds.add(gid.getGuestId());
+        Iterator<Entry<String, ArrayList<GuestId>>> hostEntries = hostGuestMap.entrySet().iterator();
+        while (hostEntries.hasNext()) {
```
Comment (awood) [inline server/src/main/java/org/candlepin/resource/HypervisorResource.java]: I would prefer to see these as for loops.  I find that to be more idiomatic to Java.

```java
for (Iterator<Entry<String, ArrayList<GuestId>>> it = hostGuestMap.entrySet().iterator(); it.hasNext(); ) {

}
```

Also I didn't notice anywhere that you were actually using the Entry object's name so why not just do `hostGuestMap.values()`?

File: server/src/main/java/org/candlepin/resource/HypervisorResource.java [hunk 5]
```diff
@@ -123,22 +125,42 @@ public HypervisorCheckInResult hypervisorUpdate(
...
+                    allGuestIds.add(guestId);
+                }
+                else {
+                    emptyGuestIdCount++;
+                    guestIds.remove();
+                }
             }
         }
+        if (emptyGuestIdCount > 0) {
+            log.warn("Found " + emptyGuestIdCount + " empty guest id(s). Ignoring.");
```
Comment (awood) [inline server/src/main/java/org/candlepin/resource/HypervisorResource.java]: Avoid string concats in log statements.

Use

```java
log.warn("Found {} empty guest id(s). Ignoring.", emptyGuestIdCount);
```

See [here](http://www.slf4j.org/manual.html)

File: server/src/main/java/org/candlepin/resource/HypervisorResource.java (hunk 1 above)
Comment (awood) [inline server/src/main/java/org/candlepin/resource/HypervisorResource.java]: List worked for me in Java 1.8.  Were there compatibility issues in 1.6?

```java

@Test
public void testRemove() {
    List<String> l = new ArrayList<String>();
    l.add("Hello");
    l.add("World");
    l.add("Foo");
    assertEquals(3, l.size());

    for (Iterator<String> x = l.iterator(); x.hasNext();) {
        String s = x.next();
        if ("Foo".equals(s)) {
            x.remove();
        }
    }

    assertEquals(2, l.size());
    assertFalse(l.contains("Foo"));
}
```

File: server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java [hunk 6]
```diff
@@ -310,4 +312,82 @@ public void checkInCreatesNoNewConsumerWhenCreateIsFalse() throws Exception {
...
+             any(Set.class))).
+             thenReturn(new VirtConsumerMap());
+        when(consumerCurator.getGuestConsumersMap(any(Owner.class), any(Set.class))).
+            thenReturn(new VirtConsumerMap());
+
+        when(ownerCurator.lookupByKey(eq(owner.getKey()))).thenReturn(owner);
+        when(principal.canAccess(eq(owner), eq(SubResource.CONSUMERS), eq(Access.CREATE))).
+            thenReturn(true);
+        when(consumerTypeCurator.lookupByLabel(
+         eq(ConsumerTypeEnum.HYPERVISOR.getLabel()))).thenReturn(hypervisorType);
```
Comment (awood) [inline server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java]: Indentation is off here.

File: server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java [hunk 7]
```diff
@@ -310,4 +312,82 @@ public void checkInCreatesNoNewConsumerWhenCreateIsFalse() throws Exception {
...
+        when(idCertService.generateIdentityCert(any(Consumer.class)))
+            .thenReturn(new IdentityCertificate());
+
+        HypervisorCheckInResult result = null;
+        try {
+            result = hypervisorResource.hypervisorUpdate(hostGuestMap,
+            principal, owner.getKey(), true);
+        }
+        catch (PersistenceException e) {
+            fail("Expected no exception");
```
Comment (awood) [inline server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java]: This isn't necessary.  The test method already throws and exception, and if an exception is thrown, JUnit marks the test as an "error" (which is actually what the result [should be](http://stackoverflow.com/questions/3425995/whats-the-difference-between-failure-and-error-in-junit) marked as)

File: server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java [hunk 8]
```diff
@@ -310,4 +312,82 @@ public void checkInCreatesNoNewConsumerWhenCreateIsFalse() throws Exception {
...
+        }
+        assertTrue(result != null);
```
Comment (awood) [inline server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java]: Use assertNotNull()

File: server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java [hunk 9]
```diff
@@ -310,4 +312,82 @@ public void checkInCreatesNoNewConsumerWhenCreateIsFalse() throws Exception {
...
+        assertTrue(result.getCreated().size() == 1);
```
Comment (awood) [inline server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java]: `assertEquals(1, result.getCreated().size())`

Using the appropriate assert yields better messaging from JUnit.

File: server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java [hunk 10]
```diff
@@ -310,4 +312,82 @@ public void checkInCreatesNoNewConsumerWhenCreateIsFalse() throws Exception {
...
+
+    @Test
+    public void ensureEmptyHypervisorIdsAreIgnored() throws Exception {
+        Owner owner = new Owner("admin");
+
+        Map<String, ArrayList<GuestId>> hostGuestMap = new HashMap<String, ArrayList<GuestId>>();
+        hostGuestMap.put("", new ArrayList(Arrays.asList(new GuestId("GUEST_A"),
+                                         new GuestId("GUEST_B"))));
+        hostGuestMap.put("HYPERVISOR_A", new ArrayList(Arrays.asList(new GuestId("GUEST_C"),
+                                                     new GuestId("GUEST_D"))));
```
Comment (awood) [inline server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java]: The indentation is inconsistent here.  Just add one additional indention on line continuations instead of spacing all the way over past the containing parenthesis.

File: server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java [hunk 11]
```diff
@@ -310,4 +312,82 @@ public void checkInCreatesNoNewConsumerWhenCreateIsFalse() throws Exception {
...
+        try {
+            result = hypervisorResource.hypervisorUpdate(hostGuestMap,
+            principal, owner.getKey(), true);
+        }
+        catch (PersistenceException e) {
+            fail("Expected no exception");
+        }
+        assertTrue(result != null);
+        assertTrue(result.getCreated().size() == 1);
+        assertTrue(!result.getCreated().toArray(new Consumer[1])[0].getHypervisorId().equals(""));
```
Comment (awood) [inline server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java]: Why convert to an array?  Does `result.getCreated().get(0)` not work?

And here again, use `assertEquals("", ...)` so you get a message like `Expected "" but was "Hello world"` instead of JUnit just throwing an `AssertionError` with zero context.

File: server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java [hunk 12]
```diff
@@ -310,4 +312,82 @@ public void checkInCreatesNoNewConsumerWhenCreateIsFalse() throws Exception {
...
+            thenReturn(true);
+        when(consumerTypeCurator.lookupByLabel(
+            eq(ConsumerTypeEnum.HYPERVISOR.getLabel()))).thenReturn(hypervisorType);
+        when(idCertService.generateIdentityCert(any(Consumer.class)))
+            .thenReturn(new IdentityCertificate());
+
+        HypervisorCheckInResult result = null;
+        try {
+            result = hypervisorResource.hypervisorUpdate(hostGuestMap,
+           principal, owner.getKey(), true);
```
Comment (awood) [inline server/src/test/java/org/candlepin/resource/HypervisorResourceTest.java]: Indentation out of whack here too

File: server/src/main/java/org/candlepin/pinsetter/tasks/HypervisorUpdateJob.java [hunk 13]
```diff
@@ -106,6 +107,49 @@ public static boolean isSchedulable(JobCurator jobCurator, JobStatus status) {
...
+                hypervisor.getHypervisorId().getHypervisorId() != null) {
+                if (hypervisor.getHypervisorId().getHypervisorId().equals("")) {
+                    // Ignore hypervisors that have a hypervisor ID of ""
+                    hypervisorsIterator.remove();
+                    emptyHypervisorIdCount++;
+                    continue;
+                }
+                incomingHosts.put(hypervisor.getHypervisorId().getHypervisorId(), hypervisor);
+                hosts.add(hypervisor.getHypervisorId().getHypervisorId());
+                if (hypervisor.getGuestIds() != null && !hypervisor.getGuestIds().isEmpty()) {
```
Comment (awood) [inline server/src/main/java/org/candlepin/pinsetter/tasks/HypervisorUpdateJob.java]: Do we really need to check if it's empty?  Won't an empty list just bomb out of the loop immediately?

**Comment by awood** [comment]
Once this is approved, if the reviewer lets me know, I will squash these commits.

**Comment by nguyenfilip** [comment]
@awood Looks good to me. Can you squash those commits?
